xbmc/favourites/FavouritesService.cpp

```diff
@@ -0,0 +1,226 @@
+    return false;
+  }
+
+  TiXmlElement *favourite = root->FirstChildElement("favourite");
+  while (favourite)
+  {
+    // format:
+    // <favourite name="Cool Video" thumb="foo.jpg">PlayMedia(c:\videos\cool_video.avi)</favourite>
+    // <favourite name="My Album" thumb="bar.tbn">ActivateWindow(MyMusic,c:\music\my album)</favourite>
+    // <favourite name="Apple Movie Trailers" thumb="path_to_thumb.png">RunScript(special://xbmc/scripts/apple movie trailers/default.py)</favourite>
+    const char *name = favourite->Attribute("name");
+    const char *thumb = favourite->Attribute("thumb");
+    if (name && favourite->FirstChild())
+    {
+      CURL url;
+      url.SetProtocol("favourites");
+      url.SetHostName(CURL::Encode(favourite->FirstChild()->Value()));
+      const std::string favURL(url.Get());
+      if (!items.Contains(favURL))
+      {
+        const CFileItemPtr item(std::make_shared<CFileItem>(name));
+        item->SetPath(favURL);
+        if (thumb)
+          item->SetArt("thumb", thumb);
+        items.Add(item);
+      }
+    }
+    favourite = favourite->NextSiblingElement("favourite");
+  }
+  return true;
+}
+
+CFavouritesService::CFavouritesService(std::string userDataFolder) : m_userDataFolder(std::move(userDataFolder))
+{
+  CFileItemList items;
+  std::string favourites = "special://xbmc/system/favourites.xml";
+  if(XFILE::CFile::Exists(favourites))
+    LoadFromFile(favourites, m_favourites);
+  else
+    CLog::Log(LOGDEBUG, "CFavourites::Load - no system favourites found, skipping");
+
+  favourites = URIUtils::AddFileToFolder(m_userDataFolder, "favourites.xml");
+  if(XFILE::CFile::Exists(favourites))
+    LoadFromFile(favourites, m_favourites);
+  else
+    CLog::Log(LOGDEBUG, "CFavourites::Load - no userdata favourites found, skipping");
+}
+
+bool CFavouritesService::Persist()
+{
+  CXBMCTinyXML doc;
+  TiXmlElement xmlRootElement("favourites");
+  TiXmlNode *rootNode = doc.InsertEndChild(xmlRootElement);
+  if (!rootNode)
+    return false;
+
+  for (const auto& item : m_favourites)
+  {
+    TiXmlElement favNode("favourite");
+    favNode.SetAttribute("name", item->GetLabel().c_str());
+    if (item->HasArt("thumb"))
+      favNode.SetAttribute("thumb", item->GetArt("thumb").c_str());
+
+    const CURL url(item->GetPath());
+    TiXmlText execute(CURL::Decode(url.GetHostName()));
+    favNode.InsertEndChild(execute);
+    rootNode->InsertEndChild(favNode);
+  }
+
+  auto path = URIUtils::AddFileToFolder(m_userDataFolder, "favourites.xml");
+  return doc.SaveFile(path);
+}
+
+bool CFavouritesService::Save(const CFileItemList& items)
+{
+  {
+    CSingleLock lock(m_criticalSection);
+    m_favourites.Clear();
+    m_favourites.Copy(items);
+    Persist();
+  }
+  OnUpdated();
+  return true;
+}
+
+void CFavouritesService::OnUpdated()
+{
+  ANNOUNCEMENT::CAnnouncementManager::GetInstance().Announce(ANNOUNCEMENT::GUI, "xbmc", "OnFavouritesUpdated");
```

Comment:
I'll rather remove the whole dependency on announcement manager (and I will, in a followup pr as this is moved code)